### ISSUE TITLE
Wrong raw string with one backslash in it

### ISSUE THREAD

**Issue by hpytree** [python/cpython#93314]
Please look at this:

```
Python 3.10.0 (tags/v3.10.0:b494f59, Oct  4 2021, 19:00:18) [MSC v.1929 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> r'\'
  File "<stdin>", line 1
    r'\'
    ^
SyntaxError: unterminated string literal (detected at line 1)
>>> r'\s'
'\\s'
>>> r'\\'
'\\\\'
>>>
```

As you see, if there is only one backslash in the raw string, python will raise a SyntaxError. Please slove it. Thank you very much!

**Comment by mdickinson** [comment]
Thanks for the report. This is intentional and documented. See https://docs.python.org/3/reference/lexical_analysis.html?highlight=raw%20string#string-and-bytes-literals, and especially this bit:

> Specifically, a raw literal cannot end in a single backslash (since the backslash would escape the following quote character).

**Comment by stevendaprano** [comment]
[This is a FAQ](https://docs.python.org/3/faq/design.html#id26). I don't know if it can be fixed. See discussion [on Python-Dev](https://mail.python.org/archives/list/python-dev@python.org/message/A437MSH3QO6CG2JIZHDTEDVUZZ2TCRYI/).

**Comment by stevendaprano** [comment]
@mdickinson That explanation in the docs has never really made sense to me. In a raw string, the backslash doesn't escape any other character, it always stands for itself. So why would it suddenly switch to escaping the next character if and only if it is the last character in the string? What possible purpose does that serve?

My understanding is that it was a limitation of the old LL(1) parser that would have been too painful to fix, but with the new PEG parser perhaps we can revisit it.

**Comment by terryjreedy** [comment]
For anyone puzzled, Mark's expression is 3 string literals implicitly joined.  Adding 2 explicit `'  +  '`s to join them:
```
>>> r"Stuff \" + r'\' + r'\""  +  ' + r"'  +  "          "
'Stuff \\" + r\'\\\' + r\'\\" + r"          '
```

**Comment by ericvsmith** [comment]
> My understanding is that it was a limitation of the old LL(1) parser that would have been too painful to fix, but with the new PEG parser perhaps we can revisit it.

This is a tokenizer (lexer) issue, not a parser issue: https://github.com/python/cpython/blob/main/Parser/tokenizer.c#L1900

I agree with @mdickinson: there's no way to change this without breaking existing code.

**Comment by tlandschoff-scale** [comment]
tl;dr: IMHO at least the documentation needs fixing.

I use Python since 1999 and professionally since sometime in 2009. I had the impression that I am quite fluent in the language and just ran into this (by trying to replace backslashes in paths). So far I thought that backslashes don't escape anything in raw strings. And the [documentation](https://docs.python.org/3.10/reference/lexical_analysis.html#string-and-bytes-literals) clearly states:

> Both string and bytes literals may optionally be prefixed with a letter 'r' or 'R'; such strings are called raw strings and treat backslashes as literal characters.

I checked by throwing a few examples at the REPL and indeed this seems to be true except at string ending. This did not make any sense to me, so I looked for a bug report and found this one.

Only by the link in this ticket did I find the section about one screen height down from the definition of `raw strings` mentioning this peculiarity. I bet I am not the only one to be surprised by this. The minimal fix for this would IMHO be to change the sentence quoted above to read

> Both string and bytes literals may optionally be prefixed with a letter 'r' or 'R'; such strings are called raw strings and treat backslashes as literal characters *except at end of string* (see below).

And of course this can't be fixed without breaking existing code. But I think language design should strive to eliminate special cases so using backslash quote at the end of a raw string could get deprecated for now. In a few years it can get forbidden and finally backslashes at end of string could be treated like elsewhere in the string.
